File: BuySell1.py
'''
Solution:
1.  Only 1 transaction is allowed, and so keep updating the minimum buying price and potential profit at each instant
    with the available minimum buying price at that instant.

Time Complexity:    O(N)
Space Complexity:   O(1)

--- Passed all testcases successfully on leetcode.
'''


class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        #   edge case checks
        if (prices == None or len(prices) == 0):
            return 0

        #   initializations
        buyingPrice = float('inf')
        profit = 0

        #   keep updating the potential Buying Price and corresponding Profits at each instant.
        for price in prices:
            buyingPrice = min(buyingPrice, price)
            profit = max(profit, price - buyingPrice)

        #   return the profit.
        return profit

File: BuySell1.java
/*
TC: O(n)
SC:O(1)
*/
class Solution {
    public int maxProfit(int[] prices) {
        
        //edge case
        if(prices.length == 0 || prices == null)return 0;
        int minPrice = Integer.MAX_VALUE;
        
        //print the prices
        int price1 = 0;
        int price2 = 0;
        int[]index = new int[2];
        int profit = 0;
        for(int i = 0; i < prices.length;i++){
            
           // minPrice = Math.min(prices[i], minPrice);
           if(minPrice > prices[i]){
               minPrice = prices[i];
               price1 = prices[i]; 
               index[0] = i;
           }
            //price2 = prices[i];
            //profit = Math.max(profit, prices[i]-minPrice);
            if(profit < prices[i]-minPrice){
                profit = prices[i]-minPrice;
                price2 = prices[i];
                index[1] = i;
            }
        }
        
        System.out.println(price1);
        System.out.println(price2);
        System.out.println(Arrays.toString(index));
        return profit;
    }
}